File: server/src/main/java/me/ajfleming/tworoomsio/timer/RoundTimer.java
package me.ajfleming.tworoomsio.timer;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;
import lombok.Data;

@Data
public class RoundTimer {

  private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
  private final long initialUnits;
  private final TimeUnit unitType;
  private final TimerTrigger onTick;
  private final TimerTrigger onEnd;
  private boolean timerRunning;
  private long unitsLeft;
  private ScheduledFuture<?> refreshTask;

  public RoundTimer(int unitsLeft, TimeUnit unitType, TimerTrigger onTick, TimerTrigger onEnd) {
    this.timerRunning = false;
    this.initialUnits = unitsLeft;
    this.unitsLeft = unitsLeft;
    this.unitType = unitType;
    this.onTick = onTick;
    this.onEnd = onEnd;
  }

  public void start() {
    if (unitsLeft > 0) {
      timerRunning = true;
      if (refreshTask == null) {
        final Runnable refresh = () -> {
          unitsLeft--;
          onTick.trigger(this);
          if (unitsLeft < 1) {
            onEnd.trigger(this);
            refreshTask.cancel(true);
          }
        };
        refreshTask = scheduler.scheduleAtFixedRate(refresh, 0, 1, unitType);
      }
    }
  }

  public void stop() {
    if (refreshTask != null) {
      refreshTask.cancel(true);
      timerRunning = false;
      onTick.trigger(this);
      refreshTask = null;
    }
  }
}

File: kotlin_server/src/main/kotlin/me/ajfleming/tworoomsio/timer/RoundTimer.kt
package me.ajfleming.tworoomsio.timer

import com.corundumstudio.socketio.SocketIOServer
import me.ajfleming.tworoomsio.model.Game
import java.util.*
import java.util.concurrent.*
import kotlin.concurrent.fixedRateTimer

class RoundTimer(
    val roundTimerName: String,
    var timer: Timer? = null,
    val onTick: TimerTrigger,
    val onEnd: TimerTrigger,
    var timerRunning: Boolean = false,
    var unitsLeft: Long
    ) {

    fun start() {
        if(unitsLeft > 0) {
            timerRunning = true;
            timer = fixedRateTimer(name = roundTimerName, initialDelay = 1000, period = 1000) {
                unitsLeft--
                onTickTrigger()
                if( unitsLeft < 1 ) {
                    onEndTrigger()
                    timer?.cancel()
                }
            }
        }
    }

    fun stop() {
        timer?.cancel()
        timerRunning = false
    }

    fun onTickTrigger() = onTick.trigger(this)
    fun onEndTrigger() = onEnd.trigger(this)

    companion object {
        fun setupTimer(initialTime: Long, game: Game, server: SocketIOServer): RoundTimer {
            return RoundTimer(
                roundTimerName = "${game.id}-timer",
                onTick = onTimerTick(server, game),
                onEnd = onTimerEnd(server,game),
                unitsLeft = initialTime
            )
        }

        fun onTimerTick(server: SocketIOServer, game: Game): TimerTrigger {
            return TimerTrigger {
                server.getRoomOperations(game.channelName).sendEvent("TIMER_UPDATE", it)
            }
        }

        fun onTimerEnd(server: SocketIOServer, game: Game): TimerTrigger {
            return TimerTrigger {
                server.getRoomOperations(game.channelName).sendEvent("TIMER_UPDATE", it)
            }
        }
    }
}